tensordict/prototype/tensorclass.py

```diff
@@ -27,6 +28,8 @@
 
 # For __future__.annotations, we keep a dict of str -> class to call the class based on the string
 CLASSES_DICT = {}
+# For nested usecase, we keep dict to save key,value pairs of non-tensor data
+TENSOR_CLASS_NON_TENSOR_DATA = collections.defaultdict(dict)
```

Comment:
@vmoens I had a chat with Alessandro and Tom offline. Alessandro suggested to consider "self" member instead of a global dict. I have tried this today. Looks like the test case behavior you mentioned here may not work. https://github.com/pytorch-labs/tensordict/pull/157#pullrequestreview-1253808373. What are your thoughts? 
CC: @tcbegley @apbard

Reply:
I agree that the global variable is not ideal, thanks @apbard for the suggestion.
What about this:
In a tensorclass, if a store field is a tensorclass, we keep the nested tensordict structure as it is now, but we also store a reference to the stored tensorclass. Very roughly put:
```python
@tesorclass 
class MyDataNested:
    a_tensor: torch.Tensor
    a_string: str

@tesorclass 
class MyDataGlob:
    inner_object: MyDataNested

data_inner = MyDataNested(torch.randn(3), "hello world")
data_outer = MyDataGlob(data_inner)

# from the outside, everything seems fine
assert data_outer.data_inner is data_inner

# we can look at the private attributes now
assert data_outer.tensordict["data_inner"] is data_inner.tensordict
assert data_outer._inner_classes["data_inner"] is data_inner

# this will: (1) look at data_inner, and realize that it is a tensoclass (not a tensor or tensordict), (2) access it from `_inner_classes` instead (3) assign the string to the inner class
data_outer.data_inner.a_string = "to infinity and beyond"
```

The need to keep 2 copies (one of the tensorclass, one of the tensordict) is to keep all the functionality from tensordict. This also means that all methods such as `__getitem__` etc should keep the `_inner_classes` dictionary updated.

@tcbegley @apbard wdyt? Am I overthinking it? Is there a simpler solution (that does not involve rewriting the entire tensordict for dataclasses)?

Reply:
@vmoens my suggestion was quite similar with a slight variation in terms of semantics.
Each tensorclass internally will have 2 private members for storage:
self.tensordata (aka a tensordict) and self.otherdata (aka a standard dict). The former will store the tensor data as current implementation, the other all other objects including nested class  
then the get and set methods will redirect stuff accordingly.

p.s we could potentially introduce a third dict to map keys to type for faster indexing and to faster sanity checks (avoiding double key)...but needs benchmarking